source/common/router/scoped_rds.cc

```diff
@@ -100,72 +107,220 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
   });
 }
 
+ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProviderHelper(
+    ScopedRdsConfigSubscription& parent, std::string scope_name,
+    envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
+    Init::Manager& init_manager)
+    : parent_(parent), scope_name_(scope_name),
+      route_provider_(static_cast<RdsRouteConfigProviderImpl*>(
+          parent_.route_config_provider_manager_
+              .createRdsRouteConfigProvider(rds, parent_.factory_context_, parent_.stat_prefix_,
+                                            init_manager)
+              .release())),
+      rds_update_callback_handle_(route_provider_->subscription().addUpdateCallback([this]() {
+        // Subscribe to RDS update.
+        parent_.onRdsConfigUpdate(scope_name_, route_provider_->subscription());
+      })) {}
+
+bool ScopedRdsConfigSubscription::addOrUpdateScopes(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& resources,
+    Init::Manager& init_manager, const std::string& version_info,
+    std::vector<std::string>& exception_msgs) {
+  bool any_applied = false;
+  envoy::config::filter::network::http_connection_manager::v2::Rds rds;
+  rds.mutable_config_source()->MergeFrom(rds_config_source_);
+  absl::flat_hash_set<std::string> unique_resource_names;
+  for (const auto& resource : resources) {
+    envoy::api::v2::ScopedRouteConfiguration scoped_route_config;
+    try {
+      scoped_route_config = MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(
+          resource.resource(), validation_visitor_);
+      MessageUtil::validate(scoped_route_config);
+      const std::string scope_name = scoped_route_config.name();
+      if (!unique_resource_names.insert(scope_name).second) {
+        throw EnvoyException(
+            fmt::format("duplicate scoped route configuration '{}' found", scope_name));
+      }
+      // TODO(stevenzzz): Creating a new RdsRouteConfigProvider likely expensive, migrate RDS to
+      // config-provider-framework to make it light weight.
+      rds.set_route_config_name(scoped_route_config.route_configuration_name());
+      // Delete previous route provider if any.
```

Comment:
Does this comment belong lower down? There's no delete here.

Reply:
moved.  Thanks